File: Problems/C++/Easy/Linked List/21. Merge Two Sorted Lists.cpp
/*
Runtime: 4 ms, faster than 94.45% of C++ online submissions for Merge Two Sorted Lists.
Memory Usage: 15.1 MB, less than 9.05% of C++ online submissions for Merge Two Sorted Lists.
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode* dummyNode = new ListNode(0);
        ListNode* head = dummyNode;
        while(l1!=nullptr || l2!=nullptr){
            ListNode* node = new ListNode();
            if(l1==nullptr){
                node->val = l2->val;
                l2 = l2->next;
                head->next = node;
                head = head->next;
                continue;
            }else if(l2==nullptr){
                node->val = l1->val;
                l1 = l1->next;
                head->next = node;
                head = head->next;
                continue;
            }
            if(l1->val<l2->val){
                node->val = l1->val;
                l1 = l1->next;
            }else{
                node->val = l2->val;
                l2 = l2->next;
            }
            head->next = node;
            head = head->next;
        }
        return dummyNode->next;
    }
};

File: Problems/Python/Easy/Linked List/21. Merge Two Sorted Lists.py
'''
Runtime: 34 ms, faster than 94.97% of Python3 online submissions for Merge Two Sorted Lists.
Memory Usage: 13.9 MB, less than 82.94% of Python3 online submissions for Merge Two Sorted Lists.
'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummyHead = ListNode(0);
        tmp = dummyHead;
        while list1 and list2:
            if list1.val > list2.val:
                tmp.next = ListNode(list2.val);
                list2 = list2.next;
            else:
                tmp.next = ListNode(list1.val);
                list1 = list1.next;
            tmp = tmp.next;
        
        while list1:
            tmp.next = ListNode(list1.val);
            tmp = tmp.next;
            list1 = list1.next;
        while list2:
            tmp.next = ListNode(list2.val);
            tmp = tmp.next;
            list2 = list2.next;
        
        return dummyHead.next;